engine/badgerengine/store.go

```diff
@@ -16,10 +16,17 @@ type Store struct {
 	name     string
 }
 
+// build a long key for each key of a store
+// in the form: storePrefix + <sep> + 0 + key.
+// the 0 is used to separate the actual key
+// from the rest of the prexix and to ensure
+// we can quickly access the latest key of the store
+// by replacing 0 by anything bigger.
```

Comment:
I am not sure to understand the last part, could you elaborate please?

Reply:
Some context: The Badger engine stores data from all the stores in the same namespace. There is no concept of separate store/bucket in Badger, so we need to namespace them ourselves, using prefixes.

The role function of this function is to generate a badger key for each key of our stores:

```
// in Genji
...
store.Put([]byte("a"), someDocument)

// is equivalent to this in Badger
k := buildKey(somePrefix, []byte("a"))
```

This function generates a key in a way that makes it simple for us the iterator to iterate over the keys of a particular store.
Long story short, whenever we want to use an Iterator to jump to the **last** key of the store, we want to do this in one step, without having to iterate first over all the keys of the store.

To do this, we use a trick. Here is what we need to do without that trick:
```go
const separator = "|"

// this function is a simplified version of how it CURRENTLY works
func buildDumbKey(storeName string, key string) string {
  return storeName + separator + key
}

buildDumbKey("foo", "bar") // foo|bar
buildDumbKey("foo", "aar") // foo|aar
buildDumbKey("foo", "b") // foo|b
buildDumbKey("coo", "bar") // coo|bar

// the badger keys now looks like this
coo|bar
foo|aar
foo|b
foo|bar
/*

// how do we get the last key of store foo?
// Currently, we try a trick but its bugged, without this trick we'd have to do this:

// create a NON reverse cursor
it := newCursor()

// iterate until the end of the prefix foo
for it.Seek("foo"); i.ValidForPrefix("foo"); it.Next() {
}
// we reached the end, we need to go back once to place the cursor on "foo|bar"
last := it.Prev()
*/

Now here is a simplified version of that trick
```go
// this function is a simplified version of how this PR works
func buildSmartKey(storeName string, key string) string {
 // note that '0' is the byte 0 in this pseudo code, not the character 0
  return storeName + separator + byte(0) + key
}

buildSmartKey("foo", "bar") // foo|0bar
buildSmartKey("foo", "aar") // foo|0aar
buildSmartKey("foo", "b") // foo|0b
buildSmartKey("coo", "bar") // coo|0bar

// to jump to the end

// create a reverse cursor this time
it := newCursor(reverse: true)

// this will seek for foo" + byte(1), which is higher than ANY key in the foo store, because 1 is bigger than 0
// I chose 1 here, but it could be any byte greater than 0 (which means 0 < theChosenByte < 256).
// We know that this key doesn't exists because no buildSmartKey uses 0 all the time, but thanks to this the
// cursor places us at the last key anyway. Why? Because when Seek doesn't find the key you're looking for, it moves to the next one, following the direction of the iterator.
it.Seek("foo" + byte(1))
```